struct.c

```diff
@@ -1663,6 +2160,30 @@ InitVM_Struct(void)
 
     rb_define_method(rb_cStruct, "deconstruct", rb_struct_to_a, 0);
     rb_define_method(rb_cStruct, "deconstruct_keys", rb_struct_deconstruct_keys, 1);
+
+    rb_cData = rb_define_class("Data", rb_cObject);
+
+    rb_undef_method(CLASS_OF(rb_cData), "new");
+    rb_undef_alloc_func(rb_cData);
+    rb_define_singleton_method(rb_cData, "define", rb_data_s_def, -1);
+
+    rb_define_singleton_method(rb_cData, "members", rb_data_s_members_m, 0);
+
+    rb_define_method(rb_cData, "initialize", rb_data_initialize_m, -1);
+    rb_define_method(rb_cData, "initialize_copy", rb_struct_init_copy, 1);
+
+    rb_define_method(rb_cData, "==", rb_data_equal, 1);
+    rb_define_method(rb_cData, "eql?", rb_data_eql, 1);
+    rb_define_method(rb_cData, "hash", rb_data_hash, 0);
+
+    rb_define_method(rb_cData, "inspect", rb_data_inspect, 0);
+    rb_define_alias(rb_cData,  "to_s", "inspect");
+    rb_define_method(rb_cData, "to_h", rb_data_to_h, 0);
+
+    rb_define_method(rb_cData, "members", rb_data_members_m, 0);
+
+    rb_define_method(rb_cData, "deconstruct", rb_data_deconstruct, 0);
+    rb_define_method(rb_cData, "deconstruct_keys", rb_data_deconstruct_keys, 1);
```

Comment:
I think all these instance methods should be defined directly on the generated class (and not on Data) to allow future optimizations by specializing them for the members, i.e., https://bugs.ruby-lang.org/issues/16122#note-69 and as I mentioned in an earlier comment on this PR.
Otherwise such optimizations would have to create a new Module which is extra footprint and would be visible in `ancestors`.

Could you make a PR to adjust that?

Reply:
@zverok It's unfortunate you didn't replied to this before the 3.2 release (maybe you didn't see it?).
The cost is we'll need to create an extra Module per Data subclass for optimization.
Hopefully that turns out to not be too much of a footprint drawback.

Given a pattern like
```ruby
Measure = Data.define(:amount, :unit) do
  def initialize(amount:, unit: '-none-') = super
end
```
as used in https://bugs.ruby-lang.org/issues/16122#note-68

This actually only works if the `initialize` defined by `Data.define` is defined on Data and not on the Data subclass (named `Measure` here), otherwise the custom `initialize` would just overwrite it and `super` wouldn't work.

So for optimizing Data methods we will need to do the same trick as for Struct, to include a Module between the subclass and `Data` and define optimized methods there, so e.g., keyword arguments are dealt much more efficiently than the generic logic in `Data#initialize`.

Reply:
> It's unfortunate you didn't replied to this before the 3.2 release (maybe you didn't see it?).

I didn't.
TBH, communicating via GitHub comment in already closed PR is not the most reliable way to reach me :) 

I am receiving GH mentions about everything that happens in ruby/ruby repo, but since Feb 24, I didn't have resource to read through them, so I just recently marked all thousands of them as "read" and started to follow the development anew... 

This is not to say I am guaranteed to see mentions unless they are in PRs I am explicitly checking. Not the best mode, probably, but I have no resource to rethink it either :shrug: I am much more accessible on Twitter, email, or in new tickets on the bug tracker, which I also follow but more rigorously. 

Also, considering... well, everything, ... I am not guaranteed to be reachable at all, so I believe any discussions about `Data` or other features I've introduced are better to maintain on the tracker.

Which, I think, is the best place to discuss the problem you raised, because I don't have a strong opinion on this. Though, I would maintain that the "`super` should refer to the default implementation" is a useful rule for any method, not only `initialize`. But how to find a compromise between this rule and efficiency, that I am not sure.